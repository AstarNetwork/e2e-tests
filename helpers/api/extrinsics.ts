import { AccountInfo } from '@polkadot/types/interfaces'
import { ApiPromise } from '@polkadot/api'
import { SubmittableExtrinsic } from '@polkadot/api-base/types'

export const balance = async (api: ApiPromise, address: string) => {
  const account = await api.query.system.account<AccountInfo>(address)
  return account.data.toJSON()
}

export const xTokensForRelayChain = (api: ApiPromise, amount: string, address: Uint8Array) => {
  return api.tx.xTokens.transfer(
    {
      Token: 'KSM',
    },
    amount,
    {
      V1: {
        parents: 1,
        interior: {
          X1: {
            AccountId32: {
              network: 'Any',
              id: address,
            },
          },
        },
      },
    },
    'Unlimited',
  )
}

export const xTokensForParaChain = (
  api: ApiPromise,
  token: object,
  parachainId: string,
  amount: string,
  address: Uint8Array,
) => {
  return api.tx.xTokens.transfer(
    token,
    amount,
    {
      V1: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: parachainId,
            },
            {
              AccountId32: {
                network: 'Any',
                id: address,
              },
            },
          ],
        },
      },
    },
    'Unlimited',
  )
}

export const xTokens = (
  api: ApiPromise,
  isRelayChain: boolean,
  parachainId: string,
  token: object,
  amount: string | bigint,
  address: Uint8Array,
) => {
  const multiLocation = isRelayChain
    ? { X1: { AccountId32: { network: 'Any', id: address } } }
    : {
        X2: [
          {
            Parachain: parachainId,
          },
          {
            AccountId32: {
              network: 'Any',
              id: address,
            },
          },
        ],
      }

  return api.tx.xTokens.transfer(
    token,
    amount,
    {
      V1: {
        parents: 1,
        interior: multiLocation,
      },
    },
    'Unlimited',
  )
}

export const xTokensV3 = (
  api: ApiPromise,
  isRelayChain: boolean,
  parachainId: string,
  token: object,
  amount: string | bigint,
  address: Uint8Array,
) => {
  const multiLocation = isRelayChain
    ? {
        X1: {
          AccountId32: {
            id: address,
          },
        },
      }
    : {
        X2: [
          {
            Parachain: parachainId,
          },
          {
            AccountId32: {
              id: address,
            },
          },
        ],
      }

  return api.tx.xTokens.transfer(
    token,
    amount,
    {
      V3: {
        parents: 1,
        interior: multiLocation,
      },
    } as any,
    'Unlimited',
  )
}

export const relayChainV3limitedReserveTransferAssets = (
  api: ApiPromise,
  parachainId: string,
  amount: string,
  address: Uint8Array,
) => {
  return api.tx.xcmPallet.limitedReserveTransferAssets(
    {
      V3: {
        parents: 0,
        interior: {
          X1: { Parachain: parachainId },
        },
      },
    },
    {
      V3: {
        parents: 0,
        interior: {
          X1: {
            AccountId32: {
              id: address,
            },
          },
        },
      },
    },
    {
      V3: [
        {
          id: { Concrete: { parents: 0, interior: 'Here' } },
          fun: { Fungible: amount },
        },
      ],
    },
    0,
    'Unlimited',
  )
}

export const xTokensTransferMulticurrencies = (
  api: ApiPromise,
  foreignAssetId: string,
  amount: string,
  parachainId: string,
  address: Uint8Array,
) => {
  return api.tx.xTokens.transferMulticurrencies(
    [
      [
        {
          ForeignAsset: foreignAssetId,
        },
        amount,
      ],
      [
        {
          Token: 'KSM',
        },
        '16000000000',
      ],
    ],
    '1',
    {
      V1: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: parachainId,
            },
            {
              AccountId32: {
                network: 'Any',
                id: address,
              },
            },
          ],
        },
      },
    },
    'Unlimited',
  )
}

export const xTokensTransferTransferMultiasset = (
  api: ApiPromise,
  interior: any[],
  amount: string,
  parachainId: string,
  address: Uint8Array,
) => {
  return api.tx.xTokens.transferMultiasset(
    {
      V1: {
        fun: {
          Fungible: amount,
        },
        id: {
          Concrete: {
            parents: 1,
            interior: {
              X3: interior,
            },
          },
        },
      },
    },
    {
      V1: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: parachainId,
            },
            {
              AccountId32: {
                network: 'Any',
                id: address,
              },
            },
          ],
        },
      },
    },
    'Unlimited',
  )
}

export const xTokensTransferTransferMultiassetV3 = (
  api: ApiPromise,
  interior: any[],
  amount: string,
  parachainId: string,
  address: Uint8Array,
) => {
  return api.tx.xTokens.transferMultiasset(
    {
      V3: {
        fun: {
          Fungible: amount,
        },
        id: {
          Concrete: {
            parents: 1,
            interior: {
              X3: interior,
            },
          },
        },
      },
    },
    {
      V3: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: parachainId,
            },
            {
              AccountId32: {
                id: address,
              },
            },
          ],
        },
      },
    } as any,
    'Unlimited',
  )
}

export const xTokensTransferMulticurrenciesV3 = (
  api: ApiPromise,
  foreignAssetId: string,
  amount: string,
  parachainId: string,
  address: Uint8Array,
) => {
  return api.tx.xTokens.transferMulticurrencies(
    [
      [
        {
          ForeignAsset: foreignAssetId,
        },
        amount,
      ],
      [
        {
          Token: 'KSM',
        },
        '16000000000',
      ],
    ],
    '1',
    {
      V3: {
        parents: 1,
        interior: {
          X2: [
            {
              Parachain: parachainId,
            },
            {
              AccountId32: {
                id: address,
              },
            },
          ],
        },
      },
    } as any,
    'Unlimited',
  )
}

export const sudo = (api: ApiPromise, call: SubmittableExtrinsic<'promise'>) => {
  return api.tx.sudo.sudo(call)
}
