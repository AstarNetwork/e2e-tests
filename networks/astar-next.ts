import { Config } from './types'

import astarConfig, { Vars } from './astar'

export default {
  ...astarConfig,
  polkadot: {
    ...astarConfig.polkadot,
    name: 'astarNext' as const,
  },
  kusama: {
    ...astarConfig.kusama,
    name: 'shidenNext' as const,
  },
  config: (opt) => ({
    ...astarConfig.config(opt),
    options: {
      wasmOverride: {
        polkadot: './wasm/astar_runtime.wasm',
        kusama: './wasm/shiden_runtime.wasm',
      }[opt.network],
    },
  }),
} satisfies Config<Vars>
