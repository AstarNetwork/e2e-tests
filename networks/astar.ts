import { Config } from './types'

export type Vars = {
  relayToken: string
}

export default {
  polkadot: {
    name: 'astar' as const,
    endpoint: 'wss://astar.api.onfinality.io/public-ws',
    relayToken: '340282366920938463463374607431768211455',
  },
  kusama: {
    name: 'shiden' as const,
    endpoint: 'wss://shiden.api.onfinality.io/public-ws',
    relayToken: '340282366920938463463374607431768211455',
  },
  config: ({ alice, relayToken }) => ({
    storages: {
      System: {
        Account: [[[alice.address], { data: { free: 1000n * 10n **18n } }]],
      },
      Assets: {
        account: [[[relayToken, alice.address], { balance: 10n ** 18n }]],
      },
      Sudo: {
        key: alice.address,
      },
      PolkadotXcm: {
        // avoid sending xcm version change notifications to makes things faster
        $removePrefix: ['versionNotifyTargets', 'versionNotifiers', 'supportedVersion'],
      },
    },
  }),
} satisfies Config<Vars>

export const astar = {
  paraId: 2006,
  paraAccount: '13YMK2eZzuFY1WZGagpYtTgbWBWGdoUD2CtrPj1mQPjY8Ldc',
  dot: 340282366920938463463374607431768211455n,
  astr: { Concrete: { parents: 0, interior: 'Here' } },
  aca: 18446744073709551616n,
  usdt: 4294969280n,
  dot_loc : { Concrete: { parents: 1, interior: 'Here' } },
  usdt_loc: { Concrete: { parents: 1, interior: { X3: [ { Parachain: 1000 }, { PalletInstance: 50 }, { GeneralIndex: 1984 }] } } },
} as const

export const shiden = {
  paraId: 2007,
  paraAccount: 'F7fq1jNy74AqkJ1DP4KqSrWtnTGtXfNVoDwFhTvvPxUvJaq',
  ksm: 340282366920938463463374607431768211455n,
  sdn: { Concrete: { parents: 0, interior: 'Here' } },
  kar: 18446744073709551618n,
  usdt: 4294969280n,
  ausd: 18446744073709551616n,
} as const
