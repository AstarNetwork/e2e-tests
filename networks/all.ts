import { Config } from './types'

import astar from './astar'
import astarNext from './astar-next'
import polkadot from './polkadot'
import statemint from './statemint'

const all = {
  polkadot,
  statemint,
  astar,
  astarNext,
} satisfies Record<string, Config>

export default all
