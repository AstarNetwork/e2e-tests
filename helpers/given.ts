import { test } from 'vitest'
import _ from 'lodash'

import { Network, NetworkNames, createContext, createNetworks } from '../networks'

class NetworkManager {
  keyring = createContext()

  async buildNetworks(names: NetworkNames[]) {
    const networkOptions = _.zipObject(_.zip(names) as any) as any as Record<NetworkNames, undefined>
    return createNetworks(networkOptions, this.keyring)
  }
}

export const given = <T extends NetworkNames>(...names: NetworkNames[]) =>
  test.extend<{
    networks: Record<T, Network>
    keyring: ReturnType<typeof createContext>
  }>({
    networks: async ({ task }, use) => {
      if (task.retry) {
        throw new Error('given cannot run with retry')
      }
      if (task.concurrent) {
        throw new Error('given cannot run in concurrent mode')
      }
      const manager = new NetworkManager()
      _.assign(task.suite, { __networkManager__: manager })
      await use(await manager.buildNetworks(names))
    },
    keyring: async ({ task }, use) => {
      // network manager is assigned above
      const manager: NetworkManager = _.get(task.suite, '__networkManager__')!
      await use(manager.keyring)
    },
  })
