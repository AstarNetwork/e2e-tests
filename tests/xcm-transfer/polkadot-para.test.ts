import { Context } from '../../networks/types'
import { query, tx } from '../../helpers/api'

import { astar } from '../../networks/astar'
import { statemint } from '../../networks/statemint'

import buildTest from './shared'

const tests = [
  // statemint -> astar
  {
    from: 'statemint',
    to: 'astar',
    name: 'USDT',
    fromStorage: ({ alice }: Context) => ({
      System: {
        account: [[[astar.paraAccount], { data: { free: 10e10 } }]],
      },
      Assets: {
        account: [[[statemint.usdtIndex, alice.address], { balance: 1e8 }]],
        asset: [[[statemint.usdtIndex], { supply: 1e8 }]],
      },
    }),
    test: {
      xcmPalletHorizontal: {
        tx: tx.xcmPallet.limitedReserveTransferAssetsV3(statemint.usdt, 1e7, tx.xcmPallet.parachainV3(1, astar.paraId)),
        fromBalance: query.assets(statemint.usdtIndex),
        toBalance: query.assets(astar.usdt),
      },
    },
  },
  // astar -> statemint
  {
    from: 'astar',
    to: 'statemint',
    name: 'USDT',
    fromStorage: ({ alice }: Context) => ({
      Assets: {
        account: [[[astar.usdt, alice.address], { balance: 1e8 }]],
      },
    }),
    test: {
      xtokenstHorizontal: {
        tx: tx.xtokens.transfer(astar.usdt, 1e7, tx.xtokens.parachainV3(statemint.paraId)),
        fromBalance: query.assets(astar.usdt),
        toBalance: query.assets(statemint.usdtIndex),
      },
    },
  },
] as const

export type TestType = (typeof tests)[number]

buildTest(tests)
