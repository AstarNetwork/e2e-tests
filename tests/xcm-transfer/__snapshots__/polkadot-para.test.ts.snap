// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`'astar' -> 'statemint' xcm transfer 'USDT' > xtokens transfer > balance on from chain 1`] = `
{
  "balance": 90000000,
  "extra": null,
  "reason": {
    "consumer": null,
  },
  "status": "Liquid",
}
`;

exports[`'astar' -> 'statemint' xcm transfer 'USDT' > xtokens transfer > balance on to chain 1`] = `
{
  "balance": "(rounded 1010000000)",
  "extra": null,
  "reason": {
    "consumer": null,
  },
  "status": "Liquid",
}
`;

exports[`'astar' -> 'statemint' xcm transfer 'USDT' > xtokens transfer > from chain hrmp messages 1`] = `
[
  {
    "data": [
      "ConcatenatedVersionedXcm",
      {
        "v2": [
          {
            "withdrawAsset": [
              {
                "fun": {
                  "fungible": 10000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x2": [
                        {
                          "palletInstance": 50,
                        },
                        {
                          "generalIndex": 1984,
                        },
                      ],
                    },
                    "parents": 0,
                  },
                },
              },
            ],
          },
          {
            "clearOrigin": null,
          },
          {
            "buyExecution": {
              "fees": {
                "fun": {
                  "fungible": 10000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x2": [
                        {
                          "palletInstance": 50,
                        },
                        {
                          "generalIndex": 1984,
                        },
                      ],
                    },
                    "parents": 0,
                  },
                },
              },
              "weightLimit": {
                "unlimited": null,
              },
            },
          },
          {
            "depositAsset": {
              "assets": {
                "wild": {
                  "all": null,
                },
              },
              "beneficiary": {
                "interior": {
                  "x1": {
                    "accountId32": {
                      "id": "0x88dc3417d5058ec4b4503e0c12ea1a0a89be200fe98922423d4334014fa6b0ee",
                      "network": {
                        "any": null,
                      },
                    },
                  },
                },
                "parents": 0,
              },
              "maxAssets": 1,
            },
          },
        ],
      },
    ],
    "recipient": 1000,
  },
]
`;

exports[`'astar' -> 'statemint' xcm transfer 'USDT' > xtokens transfer > to chain xcm events 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
      "weight": {
        "proofSize": "(rounded 3600)",
        "refTime": "(rounded 310000000)",
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
]
`;

exports[`'astar' -> 'statemint' xcm transfer 'USDT' > xtokens transfer > tx events 1`] = `
[
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 10000000,
          },
          "id": {
            "Concrete": {
              "interior": {
                "X3": [
                  {
                    "Parachain": 1000,
                  },
                  {
                    "PalletInstance": 50,
                  },
                  {
                    "GeneralIndex": "(rounded 2000)",
                  },
                ],
              },
              "parents": 1,
            },
          },
        },
      ],
      "dest": {
        "interior": {
          "X2": [
            {
              "Parachain": 1000,
            },
            {
              "AccountId32": {
                "id": "(hash)",
                "network": null,
              },
            },
          ],
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 10000000,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X3": [
                {
                  "Parachain": 1000,
                },
                {
                  "PalletInstance": 50,
                },
                {
                  "GeneralIndex": "(rounded 2000)",
                },
              ],
            },
            "parents": 1,
          },
        },
      },
      "sender": "Z2kDhBaTFjDS4eyFK26rBVTsy5EQUfHxgcvsYeQ95N73hh9",
    },
    "method": "TransferredMultiAssets",
    "section": "xTokens",
  },
]
`;

exports[`'statemint' -> 'astar' xcm transfer 'USDT' > xcmPallet transfer > balance on from chain 1`] = `
{
  "balance": 90000000,
  "extra": null,
  "reason": {
    "consumer": null,
  },
  "status": "Liquid",
}
`;

exports[`'statemint' -> 'astar' xcm transfer 'USDT' > xcmPallet transfer > balance on to chain 1`] = `
{
  "balance": "(rounded 10000000)",
  "extra": null,
  "reason": {
    "sufficient": null,
  },
  "status": "Liquid",
}
`;

exports[`'statemint' -> 'astar' xcm transfer 'USDT' > xcmPallet transfer > from chain hrmp messages 1`] = `
[
  {
    "data": [
      "ConcatenatedVersionedXcm",
      {
        "v3": [
          {
            "reserveAssetDeposited": [
              {
                "fun": {
                  "fungible": 10000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x3": [
                        {
                          "parachain": 1000,
                        },
                        {
                          "palletInstance": 50,
                        },
                        {
                          "generalIndex": 1984,
                        },
                      ],
                    },
                    "parents": 1,
                  },
                },
              },
            ],
          },
          {
            "clearOrigin": null,
          },
          {
            "buyExecution": {
              "fees": {
                "fun": {
                  "fungible": 10000000,
                },
                "id": {
                  "concrete": {
                    "interior": {
                      "x3": [
                        {
                          "parachain": 1000,
                        },
                        {
                          "palletInstance": 50,
                        },
                        {
                          "generalIndex": 1984,
                        },
                      ],
                    },
                    "parents": 1,
                  },
                },
              },
              "weightLimit": {
                "unlimited": null,
              },
            },
          },
          {
            "depositAsset": {
              "assets": {
                "wild": {
                  "allCounted": 1,
                },
              },
              "beneficiary": {
                "interior": {
                  "x1": {
                    "accountId32": {
                      "id": "0x88dc3417d5058ec4b4503e0c12ea1a0a89be200fe98922423d4334014fa6b0ee",
                      "network": null,
                    },
                  },
                },
                "parents": 0,
              },
            },
          },
          {
            "setTopic": "redacted",
          },
        ],
      },
    ],
    "recipient": 2006,
  },
]
`;

exports[`'statemint' -> 'astar' xcm transfer 'USDT' > xcmPallet transfer > to chain xcm events 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
      "weight": {
        "proofSize": "(rounded 16000)",
        "refTime": 4000000000,
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
]
`;

exports[`'statemint' -> 'astar' xcm transfer 'USDT' > xcmPallet transfer > tx events 1`] = `
[
  {
    "data": [
      {
        "Complete": {
          "proofSize": "(rounded 6200)",
          "refTime": "(rounded 673000000)",
        },
      },
    ],
    "method": "Attempted",
    "section": "polkadotXcm",
  },
]
`;
